src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -6266,13 +6338,120 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
                 return this.AddError(name, ErrorCode.ERR_TypeExpected);
             }
         }
 
+        private FunctionPointerTypeSyntax ParseFunctionPointerTypeSyntax()
+        {
+            Debug.Assert(IsFunctionPointerStart());
+            var @delegate = EatToken(SyntaxKind.DelegateKeyword);
+            var asterisk = EatToken(SyntaxKind.AsteriskToken);
+            var callingConvention = CurrentToken.Kind != SyntaxKind.LessThanToken
+                ? CurrentToken.ContextualKind switch
+                {
+                    SyntaxKind.CdeclKeyword => EatContextualToken(SyntaxKind.CdeclKeyword),
+                    SyntaxKind.ManagedKeyword => EatContextualToken(SyntaxKind.ManagedKeyword),
+                    SyntaxKind.UnmanagedKeyword => EatContextualToken(SyntaxKind.UnmanagedKeyword),
+                    SyntaxKind.ThiscallKeyword => EatContextualToken(SyntaxKind.ThiscallKeyword),
+                    SyntaxKind.StdcallKeyword => EatContextualToken(SyntaxKind.StdcallKeyword),
+                    _ => parseInvalidCallingConvention(),
+                }
+                : null;
+
+            var lessThanToken = EatToken(SyntaxKind.LessThanToken);
+            var saveTerm = _termState;
+            _termState |= TerminatorState.IsEndOfFunctionPointerParameterList;
+            var types = _pool.AllocateSeparated<FunctionPointerParameterOrReturnTypeSyntax>();
+
+            try
+            {
+                while (true)
+                {
+                    var modifiers = _pool.Allocate<SyntaxToken>();
+                    try
+                    {
+                        switch (CurrentToken.Kind)
+                        {
+                            case SyntaxKind.ReadOnlyKeyword:
+                                modifiers.Add(EatToken(SyntaxKind.ReadOnlyKeyword));
+                                modifiers.Add(EatToken(SyntaxKind.RefKeyword));
+                                break;
+
+                            case SyntaxKind.RefKeyword:
+                            case SyntaxKind.OutKeyword:
+                            case SyntaxKind.InKeyword:
+                                modifiers.Add(EatToken(CurrentToken.Kind));
+                                break;
+                        }
+
+                        var parameterType = ParseTypeOrVoid();
+                        types.Add(SyntaxFactory.FunctionPointerParameterOrReturnType(modifiers, parameterType));
+
+                        if (CurrentToken.Kind == SyntaxKind.CommaToken)
+                        {
+                            types.AddSeparator(EatToken(SyntaxKind.CommaToken));
+                        }
+                        else if (skipBadFunctionPointerParameterListTokens() == PostSkipAction.Abort)
+                        {
+                            break;
+                        }
+
+                        PostSkipAction skipBadFunctionPointerParameterListTokens()
```

Comment:
ick.  can you move this below.  it gets really confusing to read interspersed normal code-flow with helper functions.

Reply:
No, it's capturing local state.

---
In reply to: [348019621](https://github.com/dotnet/roslyn/pull/39879#discussion_r348019621) [](ancestors = 348019621)

Reply:
thanks.  now i hate it more :)

Reply:
I could put a `continue` above it to make the separation clearer?

---
In reply to: [348237933](https://github.com/dotnet/roslyn/pull/39879#discussion_r348237933) [](ancestors = 348237933)

Reply:
could you instead move it, and just pass the values directly, instead of capturing them?

Reply:
Apparently I lied, the variable I was capturing is at the top level and is fine to be on the outer level.

---
In reply to: [348244719](https://github.com/dotnet/roslyn/pull/39879#discussion_r348244719) [](ancestors = 348244719)